core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java

```diff
@@ -143,6 +146,50 @@ public void onShardFailedFailure(DiscoveryNode master, TransportException e) {
         assertTrue(failure.get());
     }
 
+    public void testTimeout() throws InterruptedException {
+        final String index = "test";
+
+        clusterService.setState(stateWithStartedPrimary(index, true, randomInt(5)));
+
+        String indexUUID = clusterService.state().metaData().index(index).getIndexUUID();
+
+        AtomicBoolean progress = new AtomicBoolean();
+        assert !progress.get();
+        AtomicBoolean timedOut = new AtomicBoolean();
+        assert !timedOut.get();
+
+        TimeValue timeout = new TimeValue(3, TimeUnit.SECONDS);
+        Object monitor = new Object();
+        long start = System.nanoTime();
+        shardStateAction.shardFailed(getRandomShardRouting(index), indexUUID, "test", getSimulatedFailure(), timeout, new ShardStateAction.Listener() {
+            @Override
+            public void onShardFailedNoMaster() {
+
+            }
+
+            @Override
+            public void onShardFailedFailure(DiscoveryNode master, TransportException e) {
+                if (e instanceof ReceiveTimeoutTransportException) {
+                    timedOut.set(true);
+                }
+                synchronized (monitor) {
+                    monitor.notify();
+                }
+            }
+        });
+
+        synchronized (monitor) {
+            monitor.wait();
+        }
+        long duration = System.nanoTime() - start;
+        assertTrue(timedOut.get());
+        assertThat(duration, greaterThanOrEqualTo(TimeUnit.SECONDS.toNanos(3L)));
```

Comment:
this is dangerous. I'm not sure we can rely on this. Also testing the exact amount tests generic Transport functionality. I don't think we should do it here. Just keep it simple.